language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/FromValueGenerator.scala

```diff
@@ -37,7 +39,58 @@ private[inner] object FromValueGenerator extends StrictLogging {
       .addParameter(TypeName.get(classOf[javaapi.data.Value]), "value$")
       .addParameters(converterParams.asJava)
       .addException(classOf[IllegalArgumentException])
-      .addCode(recordValueExtractor("value$", "recordValue$"))
+      .addAnnotation(classOf[Deprecated])
+      .addJavadoc(
+        "@deprecated since Daml $L; $L",
+        "2.5.0",
+        s"use {@code valueDecoder} instead",
+      )
+
+    val fromValueParams = CodeBlock.join(
+      converterParams.map { param =>
+        CodeBlock.of("$N::apply", param)
```

Comment:
It now occurs to me that this will cause problems with patterns like `ParameterizedContractId`, which should continue to work even though deprecated.

You should be able to write a test that currently demonstrates that this throws. (You will _not_ see it throw by decoding _template data_ containing a `ParameterizedContractId`. That's because the deprecated `fromValue` for template data uses the template's decoder, which doesn't follow the `::apply` path at all to represent decoding the type param. It can only be observed directly.)

The fix I suggest:

1. Put a utility static method somewhere in the `codegen` package to convert `Function<Value, T>` to `ValueDecoder<T>`
2. _This method should be immediately deprecated_, because you should only have to call it from deprecated generated code.
3. It should override `fromContractId` with the old behavior of `new ContractId<>`; the `default` of throwing is suitable for the new methods, but will break old code.
4. Use this utility method instead of `::apply` here.

This sort of pattern should apply to all sorts of generated code, in truth. Anywhere we can turn a repetitive pattern in codegen, into hand-written, easily reviewable code, is a reliability win.